src/main/java/org/jboss/threads/EnhancedQueueExecutor.java

```diff
@@ -2226,12 +2240,14 @@ void park(EnhancedQueueExecutor enhancedQueueExecutor, long nanos) {
                 //overruns a bit (as the nano time is for thread timeout) we just spin then check
                 //to keep performance consistent between the two versions.
                 do {
-                    if (tl.nextInt(PARK_SPINS) < YIELD_FACTOR) {
-                        Thread.yield();
-                    }
                     if (unsafe.compareAndSwapInt(this, parkedOffset, STATE_UNPARKED, STATE_NORMAL)) {
                         return;
                     }
+                    if (tl.nextInt(PARK_SPINS) < YIELD_FACTOR) {
+                        Thread.yield();
+                    } else {
+                        JDKSpecific.onSpinWait();
+                    }
```

Comment:
I wonder how important the `nanoTime` call is in this case as it's ignored while we spin. Perhaps we can round the spin time to zero and park the full duration when PARK_SPINS is reached?

Reply:
Well on the one hand, ~128 spins would probably take a few microseconds at most, but on the other hand, `yield` could take an arbitrary amount of time.  Maybe we should only check the time (the second time) if we yielded.

Reply:
In that case we might as well defer the first nanoTime call until just prior to the first `yield`, that is unless the nanotime call itself has timing implications that make the spin more likely to avoid parking.

Reply:
TBH I'm not really sure of the "random yield" approach.  If we yielded, does it make sense to spin again some tiny amount of times afterwards?  Does it make sense to sometimes yield after fewer spins?  I'm suspecting the answer to both of these is "no" - maybe it makes more sense to spin N times, then yield M times, then park.  This would also simplify the elapsed time logic.

Reply:
Agreed, it's much easier for me to reason about an approach that escalates from spin into yield, and finally park based on well defined inputs.

Separately, though more concerning using the deterministic escalation approach, it's odd that we can yield then exit the loop without a cas check.

Reply:
Yeah, that's problematic and I suspect it could result in measurable degradation.  Being deterministic makes it easier to code correctly though I suspect.